camel/models/samba_model.py

```diff
@@ -0,0 +1,159 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import os
+from typing import Any, Dict, List, Optional
+
+from openai import OpenAI, Stream
+
+from camel.configs import SAMBA_API_PARAMS
+from camel.messages import OpenAIMessage
+from camel.types import ChatCompletionChunk, ModelType
+from camel.utils import (
+    BaseTokenCounter,
+    OpenAITokenCounter,
+    api_keys_required,
+)
+
+
+class SambaModel:
+    r"""SambaNova service interface."""
+
+    def __init__(
+        self,
+        model_type: ModelType,
+        model_config_dict: Dict[str, Any],
+        api_key: Optional[str] = None,
+        url: Optional[str] = None,
+        token_counter: Optional[BaseTokenCounter] = None,
+    ) -> None:
+        r"""Constructor for SambaNova backend.
+
+        Args:
+            model_type (ModelType): Model for which a backend is created.
+                Currently only support `"llama3-405b"`
+            api_key (Optional[str]): The API key for authenticating with the
+                SambaNova service. (default: :obj:`None`)
+            url (Optional[str]): The url to the SambaNova service. (default:
+                :obj:`"https://fast-api.snova.ai/v1/chat/completions"`)
+            token_counter (Optional[BaseTokenCounter]): Token counter to use
+                for the model. If not provided, `OpenAITokenCounter(ModelType.
+                GPT_3_5_TURBO)` will be used.
+        """
+        self.model_type = model_type
+        self._api_key = api_key or os.environ.get("SAMBA_API_KEY")
+        self._url = url or os.environ.get(
+            "SAMBA_API_BASE_URL",
+            "https://fast-api.snova.ai/v1/chat/completions",
+        )
+        self._token_counter = token_counter
+        self.model_config_dict = model_config_dict
+        self.check_model_config()
+
+    @property
+    def token_counter(self) -> BaseTokenCounter:
+        r"""Initialize the token counter for the model backend.
+
+        Returns:
+            BaseTokenCounter: The token counter following the model's
+                tokenization style.
+        """
+        if not self._token_counter:
+            self._token_counter = OpenAITokenCounter(ModelType.GPT_3_5_TURBO)
+        return self._token_counter
+
+    def check_model_config(self):
+        r"""Check whether the model configuration contains any
+        unexpected arguments to SambaNova API.
+
+        Raises:
+            ValueError: If the model configuration dictionary contains any
+                unexpected arguments to SambaNova API.
+        """
+        for param in self.model_config_dict:
+            if param not in SAMBA_API_PARAMS:
+                raise ValueError(
+                    f"Unexpected argument `{param}` is "
+                    "input into SambaNova model backend."
+                )
+
+    @api_keys_required("SAMBA_API_KEY")
```

Comment:
as discussed, the bug is due to api_key is passed to `ChatAgent` instead of `ModelFactory.create`, after the refactor of model part we shouldn't keep the api_key passing for `ChatAgent`, thanks for raising this issue!